Add Summarizer (standalone + node in custom pipelines + SearchSummarizationPipeline)

lalitpagaria opened the pull request:
To complete #651 

@tholor Please review, specially `generate_one_summary` option and one  TODO in `TransformersSummarizer`.

With `generate_one_summary` we can achieve both what you have [asked on original issue](https://github.com/deepset-ai/haystack/issues/651#issuecomment-737707756)

> I see two different options to implement:
> a) retrieve n docs and give 1 summary
> b) retrieve n docs and give 1 summary per doc

I am skipping tutorial and document (Readme etc) update for now. We can do in separate PR.

lalitpagaria commented on .github/workflows/ci.yml:
```diff
@@ -14,13 +14,6 @@ jobs:
     steps:
     - uses: actions/checkout@v2
       
-    - name: Configure sysctl limits for Elasticsearch
```
This is not required as we are running ElasticSearch in docker container

lalitpagaria commented on .github/workflows/ci.yml:
```diff
@@ -40,14 +33,17 @@ jobs:
     - name: Run pytest with generator/pipeline marker
       run: cd test && pytest -m "pipeline or generator"
 
+    - name: Run pytest with summarizer marker
```
Because of OOM issue

lalitpagaria commented on haystack/summarizer/transformers.py:
```diff
@@ -0,0 +1,137 @@
+            # TODO: How to fix this?
+            # Documents order is very important to produce summary.
+            # Different order of same documents produce different summary.
+            contexts = [self.padding_for_one_summary.join(contexts)]
```
Is this way to produce single summary is fine?

tholor replied:
Yes, I think it's inevitable that the generated summary depends on the order of joined docs when using these kinds of models.

lalitpagaria commented:
We need to improve CI workflow. It took 30min to complete this whole work. But then took 2-3 hours to fix CI builds because of OOM issue. Also better to keep `mypy` step as first step, it will give early feedback about formatting instead of running full tests. I am thinking of separating CIs in two stages -
1) Fast CI: mypy, fast tests
2) Slow CI: all slow tests

WDYT?

tholor commented on haystack/pipeline.py:
```diff
@@ -239,6 +240,34 @@ def run(self, query: str, filters: Optional[Dict] = None, top_k_retriever: int =
+            query=query, filters=filters, top_k_retriever=top_k_retriever, generate_single_summary=generate_single_summary
+        )
+
+        # TODO do conversion of docs to answers here?
```
We could return `Document` objects from the summarizer and format them to "answers" here ...

lalitpagaria replied:
Do you want change in this PR or in separate PR?

tholor replied:
I would do it directly in this PR to avoid any breaking changes ...

tholor reviewed (commented):
Nice work! My comments:
- I tried to make the naming of a few args and the pipeline class more intuitive.
- I updated a few doc strings
- The problem of order-dependent summaries when creating a single summary from all docs is inevitable with these type of models, so I think your implementation is good here and we should only document this unintuitive behavior very well.
- I like the general code design and only have some doubts about one design choice here: `Summarizer.predict()` returns an answer object. I wonder if it might be better to return Document objects instead. I see the summarizer rather as a flexible, general "node" within different pipelines. The summaries might be used as "answers" in the end, but I can also imagine many other workflows. If we go in that direction we could also get rid of the `query` arg that gets passed around only for the sake of putting it in the final "answer" dict again. In order to have a "drop-in replacement" for ExtractiveQA / GenerativeQA etc., we could still do a conversion from Documents -> Answers in the pipeline itself (similar to FAQPipeline). What do you think?

tholor commented:
> We need to improve CI workflow.

Absolutely. I like the suggestion of putting mypy and fast tests first, so that we save time if those fail. 

> I am thinking of separating CIs in two stages

Do you mean two sequential stages in a GitHub workflow or two parallel actions?

We also recently had a discussion, whether we want to switch to hosting our own runners to avoid these annoying OOM issues, but this might come with other disadvantages (costs, behavior in forks ...).

lalitpagaria commented:
> I tried to make the naming of a few args and the pipeline class more intuitive.
> I updated a few doc strings
> 

I checked your changes, these were good. I am not good at comments :)
Two times I changed name of the pipeline could not come up with proper name, but your suggestion is good one.

> * I like the general code design and only have some doubts about one design choice here: `Summarizer.predict()` returns an answer object. I wonder if it might be better to return Document objects instead. I see the summarizer rather as a flexible, general "node" within different pipelines. The summaries might be used as "answers" in the end, but I can also imagine many other workflows. If we go in that direction we could also get rid of the `query` arg that gets passed around only for the sake of putting it in the final "answer" dict again. In order to have a "drop-in replacement" for ExtractiveQA / GenerativeQA etc., we could still do a conversion from Documents -> Answers in the pipeline itself (similar to FAQPipeline). What do you think?

Yes. I also do not liked it but kept as a sake of continuity. If I may suggest then how about introducing one `Convertor` node. Main idea for this node would be change from one dict format to another. And user can customize it based on their convenience. I am using this in my project as payload customizer, transformation of one payload to another format. Refer these two places -
https://github.com/lalitpagaria/obsei/blob/master/obsei/sink/jira_sink.py#L16
https://github.com/lalitpagaria/obsei/blob/master/obsei/sink/jira_sink.py#L16

So mainly by default Pipeline node's output will be standard one as we have currently but user can update by adding convertor. So if they want to send data to API then can customize payload based on their need, if they want to show data in markdown they can via convertor.
WDYT?

lalitpagaria commented:
> Do you mean two sequential stages in a GitHub workflow or two parallel actions?

For now sequential is fine, I am not sure free plan allow parallel. If then then I would go for parallel. as currently we get feedback after 20 min. With parallel run we would able to get feedback of CI run in less than 10 min.

lalitpagaria commented:
@tholor is this ready to merge or is there something you like to add?

tholor commented:
I wanted to first explore again the return type (answers or documents) of `Summarizer.predict()`. I will have a look at it now...

tholor commented:
@lalitpagaria I have changed now the return type of the summarizer and added the conversion to "answers" only in the pipeline. While I think this makes a bit more sense then the previous version, I am still not super happy with it. I think we need to address this in a separate PR more generally and re-think the input/output format of nodes within a pipeline. Right now it's quite a messy dictionary containing documents, answers, params - all bundled with **kwargs. Maybe we can find a cleaner solution there. 
I also like your idea of a `Conversion` Node that could convert the output formats. I am not sure yet if it would be applicable in our case here (as the conversion from doc to answer is here quite specific for summarization) - but it would definitely be helpful in other scenarios (e.g. converting input format of query or converting from Document objects into HTML etc.). Let's tackle this in separate PRs though ... 

If you are fine with the changes, we'll be ready to merge...

lalitpagaria commented:
LGTM
Thanks 🙂

tholor commented:
The CI is currently failing with 
``` 
self = <sqlalchemy.dialects.sqlite.pysqlite.SQLiteDialect_pysqlite object at 0x7f42617e6a50>
dbapi_connection = <sqlalchemy.pool.base._ConnectionFairy object at 0x7f4261e5d450>

    def do_commit(self, dbapi_connection):
>       dbapi_connection.commit()
E       sqlalchemy.exc.OperationalError: (sqlite3.OperationalError) disk I/O error
E       (Background on this error at: http://sqlalche.me/e/13/e3q8)
```

But this is not related to this PR as we face the same issue in other PRs (e.g. #717 ). Will need to investigate this and fix the CI tomorrow.
